Add STM32L4 PLLSAI1 and PLLSAI2 support

avolmat-st opened the pull request:
Series to add PLLSAI1 and PLLSAI2 PLLs support on STM32.  This had the support in the common part and then add the STM32L4 specific bits.

Commit:
dts: bindings: stm32_clocks: add bindings for PLLSAI of STM32L4

Add description of the SAI1 and SAI2 PLLs of the stm32l4.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: arm: st: add pllsai1 node into stm32l4.dtsi

Add a disabled node describing the PLLSAI1 pll within the stm32l4.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
drivers: clock: stm32: add PLLSAI1 support (common + L4)

Add stm32 common clock handling for the SAI1 PLL.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

erwango commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -202,6 +202,20 @@
+#define STM32_PLLSAI2_Q_DIVISOR		DT_PROP_OR(DT_NODELABEL(pllsai2), div_q, 1)
+#define STM32_PLLSAI2_R_ENABLED		DT_NODE_HAS_PROP(DT_NODELABEL(pllsai2), div_r)
+#define STM32_PLLSAI2_R_DIVISOR		DT_PROP_OR(DT_NODELABEL(pllsai2), div_r, 1)
+#define STM32_PLLSAI2_DIVR_ENABLED	DT_NODE_HAS_PROP(DT_NODELABEL(pllsai2), div_divr)
```
Do we actually need this definition ?

avolmat-st replied:
This last one is about the LTDC specific divider, which is at the tail of the R output of the PLLSAI2. I wrote it in a similar way as the other, with both an ENABLED define and a second define with the value itself.
Should I do it in another way ?

erwango replied:
Point is that `ENABLED` has a meaning for outputs as PLL outputs may actually be disabled (at least on some series), so the `ENABLED` make sense in this case. Here there is already a `STM32_PLLSAI2_R_ENABLED` for output R. Fo r the divisor, we just need the value (or presence) information not an additional flag about the fact that it is enabled

erwango commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -100,7 +105,7 @@
 #define ADFSDM_SEL(val)		STM32_DOMAIN_CLOCK(val, 3, 3, CCIPR2_REG)
 #define SAI1_SEL(val)		STM32_DOMAIN_CLOCK(val, 7, 5, CCIPR2_REG)
-/* #define SAI2_SEL(val)		STM32_DOMAIN_CLOCK(val, 7, 8, CCIPR2_REG) */
+#define SAI2_SEL(val)		STM32_DOMAIN_CLOCK(val, 7, 8, CCIPR2_REG)
```
Ditto

erwango commented on dts/arm/st/l4/stm32l4.dtsi:
```diff
@@ -104,6 +104,18 @@
 			compatible = "st,stm32l4-pll-clock";
 			status = "disabled";
 		};
+
+		pllsai1: pllsai1 {
+			#clock-cells = <0>;
+			compatible = "st,stm32l4-pllsai1-clock";
+			status = "disabled";
+		};
+
+		pllsai2: pllsai2 {
+			#clock-cells = <0>;
+			compatible = "st,stm32l4-pllsai2-clock";
+			status = "disabled";
+		};
```
Not available on all stm32l4 series (412, 422 for instance), so those nodes should be added in the right .dtsi files

avolmat-st replied:
Okay

avolmat-st commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -96,7 +99,7 @@
 #define DFSDM_SEL(val)		STM32_DOMAIN_CLOCK(val, 1, 2, CCIPR2_REG)
 #define ADFSDM_SEL(val)		STM32_DOMAIN_CLOCK(val, 3, 3, CCIPR2_REG)
-/* #define SAI1_SEL(val)		STM32_DOMAIN_CLOCK(val, 7, 5, CCIPR2_REG) */
+#define SAI1_SEL(val)		STM32_DOMAIN_CLOCK(val, 7, 5, CCIPR2_REG)
```
Indeed, maybe we can simply split between the stm32l4+ series and the other. We could have just a single new header file, say include/zephyr/dt-bindings/clock/stm32l4plus_clock.h which includes the common stm32l4_clock.h and perform undef SAI1_SEL then set it to the proper field for the stm32l4+ series.
Then the related DT of those roughly 4 series (P/Q/S/R) would include the stm32l4plus-clock header while the other would stay on the stm32l4-clock.h

Commit:
include: dt-bindings: stm32_clocks: add stm32l4+ bindings header

On stm32l4+ series, the SAI1 and SAI2 clock source selection fields
are within a different register hence a different binding header
file is introduced to allow proper field usage on those platforms.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
drivers: clock: stm32: add PLLSAI2 support (common + L4)

Add stm32 common clock handling for the SAI2 PLL.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: arm: st: add pllsai2 node for stm32l4 series 7 and upper

Add a disabled node describing the PLLSAI2 pll available from
stm32l47* and upper.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: arm: st: include stm32l4plus_clock.h in stm32l4+ dtsi

Add include of the stm32l4+ specific clock bindings in
stm32l4p5.dtsi in order to let all stm32l4+ benefit from it.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

gautierg-st commented on drivers/clock_control/clock_stm32_ll_common.h:
```diff
@@ -42,6 +42,33 @@
 #define z_plli2s_r(v) LL_RCC_PLLI2SR_DIV_ ## v
 #define plli2sr(v) z_plli2s_r(v)
 
+#define z_pllsai1_m(v) LL_RCC_PLLSAI1M_DIV_ ## v
+#define pllsai1m(v) z_pllsai1_m(v)
```
For the non-L4+, `pllsai1m` should be the same as `pllm`. You can use `RCC_PLLSAI1M_DIV_1_16_SUPPORT` to discriminate

gautierg-st commented on drivers/clock_control/clock_stm32_ll_common.h:
```diff
@@ -42,6 +42,33 @@
+#define z_pllsai1_r(v) LL_RCC_PLLSAI1R_DIV_ ## v
+#define pllsai1r(v) z_pllsai1_r(v)
+
+#define z_pllsai2_m(v) LL_RCC_PLLSAI2M_DIV_ ## v
+#define pllsai2m(v) z_pllsai2_m(v)
```
For the non-L4+, `pllsai2m` should be the same as `pllm`. You can use `RCC_PLLSAI1M_DIV_1_16_SUPPORT` to discriminate

gautierg-st reviewed (changes requested):
Nice work, a few points to fix, but I wonder the usefulness of declaring new compatibles when they are so close to the main PLL (this is open to discussion).
Here is a table listing all values for all PLLs. Blank means not available. Hopefully it is complete.
```
|        | PLL  |       |      |         |         | PLLSAI1 |       |      |         |         | PLLSAI2 |       |      |         |         |
|--------|:----:|:-----:|:----:|:-------:|:-------:|:-------:|:-----:|:----:|:-------:|:-------:|:-------:|:-----:|:----:|:-------:|:-------:|
|        | M    | N     | P    | Q       | R       | M       | N     | P    | Q       | R       | M       | N     | P    | Q       | R       |
| L41-46 | 1-8  | 8-86  | 2-31 | 2/4/6/8 | 2/4/6/8 |         | 8-86  | 2-31 | 2/4/6/8 | 2/4/6/8 |         |       |      |         |         |
| L47/48 | 1-8  | 8-86  | 7/17 | 2/4/6/8 | 2/4/6/8 |         | 8-86  | 7/17 | 2/4/6/8 | 2/4/6/8 |         | 8-86  | 7/17 |         | 2/4/6/8 |
| L49/4A | 1-8  | 8-86  | 2-31 | 2/4/6/8 | 2/4/6/8 |         | 8-86  | 2-31 | 2/4/6/8 | 2/4/6/8 |         | 8-86  | 2-31 |         | 2/4/6/8 |
| L4+    | 1-16 | 8-127 | 2-31 | 2/4/6/8 | 2/4/6/8 | 1-16    | 8-127 | 2-31 | 2/4/6/8 | 2/4/6/8 | 1-16    | 8-127 | 2-31 | 2/4/6/8 | 2/4/6/8 |
```
Main differences are the presence/absence of PLLxM, and the presence of PLLSAI2DIVR for L4+ (not represented in the table).

avolmat-st commented on dts/bindings/clock/st,stm32l4-pllsai1-clock.yaml:
```diff
@@ -0,0 +1,77 @@
+  clocks:
+    required: true
+
+  div-m:
```
Is that enough to just write within the description ? My feeling was that it would be helpful to have a dedicated compatible in such case since at compile time it would already complain if a div-m is provided while this is not available on the platform. In such case here the idea would be to have a st,stm32l4plus-pllsai1-clock.yaml which include the st,stm32l4-pllsai1-clock.yaml and add the div-m.
Does that make sense ?

avolmat-st commented on dts/bindings/clock/st,stm32l4-pllsai1-clock.yaml:
```diff
@@ -0,0 +1,77 @@
+    required: true
+    description: |
+        Multiplication factor for VCO
+        Valid range: 8 - 127
```
Of, so here as well, this would require a st,stm32l4plus-pllsai1-clock.yaml to distinguish between the L4 and L4+. What do you think ?

Based on your table above, I see also that on L47/48 for PLL the P range is different from the other version of L4. How do you intend to take care of this difference on the PLL compatible side ? Currently the PLL yaml mention

  div-p:
    type: int
    description: |
        Main PLL division factor for PLLSAI3CLK
    enum:
      - 7
      - 17

avolmat-st commented on dts/bindings/clock/st,stm32l4-pllsai1-clock.yaml:
```diff
@@ -0,0 +1,77 @@
+    description: |
+        Division factor for PLLSAI_P
+    enum:
+      - 2
+      - 31
```
Same point as above. Is there another option apart from having a dedicated yaml for the L47x/L48x ?

avolmat-st commented on drivers/clock_control/clock_stm32l4_l5_wb_wl.c:
```diff
@@ -139,6 +139,72 @@ void config_pllsai1(void)
+
+#if STM32_PLLSAI2_R_ENABLED
+#if defined(RCC_CCIPR2_PLLSAI2DIVR)
+#if STM32_PLLSAI2_DIVR_ENABLED
```
Added a well the DIVR_ENABLED check in the stm32_clock_control_get_subsys_rate

gautierg-st commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -435,11 +449,35 @@
 	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(pllsai1), st_stm32l4_pllsai_clock, okay) && \
 	!DT_SAME_NODE(DT_PLL_CLOCKS_CTRL, DT_PLLSAI1_CLOCKS_CTRL)
 #error "On STM32L4 series, PLL / PLLSAI1 must have the same source"
+#elif DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(pll), st_stm32l4_pll_clock, okay) && \
```
Do not use elif here, use if/endif each time, all three conditions should be tested. 
As it is, if all PLLs are enabled, and if PLL and PLLSAI1 share the same clock source, but PLLSAI2 uses a different one, it won't be detected.

avolmat-st replied:
Indeed. Let me fix that.

gautierg-st reviewed (commented):
Sorry, I have one last comment. Otherwise it seems good, I think we covered every potential corner case.

gautierg-st reviewed (dismissed):
You have a small compliance issue left, but otherwise it looks good to me.
